src/types.py

```diff
@@ -239,67 +320,164 @@ def get_function_pointer_signature(self) -> Optional["FunctionSignature"]:
+        If `remaining_bits` is nonzero, then there was *not* a field at the exact
+        offset provided; the returned field is at `(offset_bits - remaining_bits)`.
+        """
+        NO_MATCHING_FIELD: Type._GetFieldResult = (None, Type.any(), None)
+
+        if self.is_array():
+            data = self.data()
+            assert data.ptr_to is not None and data.array_dim is not None
+
+            size = data.ptr_to.get_size_bits()
+            assert size is not None
+            index, remaining_bits = divmod(offset_bits, size)
+            if index >= data.array_dim:
+                return NO_MATCHING_FIELD
+            subpath, subtype, sub_remainder_bits = data.ptr_to.get_field(
+                remaining_bits, target_size_bits=target_size_bits
+            )
+            if subpath is None or sub_remainder_bits is None:
+                return NO_MATCHING_FIELD
+            subpath.insert(0, index)
+            return subpath, subtype, sub_remainder_bits
+
+        if self.is_struct():
+            data = self.data()
+            assert data.struct is not None
+            possible_fields = data.struct.fields_at_offset(offset_bits)
+            if not possible_fields:
+                return NO_MATCHING_FIELD
+            possible_results: List[Type._GetFieldResult] = []
+            if target_size_bits is None or target_size_bits == self.get_size_bits():
+                possible_results.append(([], self, offset_bits))
+            for field in possible_fields:
+                inner_offset_bits = offset_bits - field.offset_bits
+                subpath, subtype, sub_remainder_bits = field.type.get_field(
+                    inner_offset_bits, target_size_bits=target_size_bits
+                )
+                if subpath is None or sub_remainder_bits is None:
+                    continue
+                subpath.insert(0, field.name)
+                possible_results.append((subpath, subtype, sub_remainder_bits))
+                if (
+                    target_size_bits is not None
+                    and target_size_bits == subtype.get_size_bits()
+                ):
+                    return possible_results[-1]
+            zero_offset_results = [r for r in possible_results if r[2] == 0]
+            if zero_offset_results:
+                return zero_offset_results[0]
+            if possible_results:
+                return possible_results[0]
+
+        if offset_bits == 0 and (
+            target_size_bits is None or target_size_bits == self.get_size_bits()
+        ):
+            # We might as well take a pointer to the whole struct/array
+            return [], self, 0
+
+        return NO_MATCHING_FIELD
+
+    def get_deref_field(
+        self, offset_bits: int, *, target_size_bits: Optional[int]
+    ) -> _GetFieldResult:
+        """
+        Similar to `.get_field()`, but treat self as a pointer (or array) and find
+        the field in the pointer's target.
```

Comment:
Yeah you're totally right. The whole math around `index` isn't needed, because `self` should always be a pointer (never an array). 

I should be able to shrink this function a lot. 

(`get_deref_field()` is the rough equivalent of the old `get_field()` because it does the dereference, whereas the new `get_field()` does not dereference and so is used to traverse the structs/arrays)